File: c++/baekjoon/1932.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <stack>
#include <queue>
#include <math.h>
using namespace std;
int arr[501][501];
int t[501][501];
int main()
{
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    int n;
    cin >>n;
    int tmp = 1;
    for(int i=0;i<n;i++){
        for(int j =0;j<tmp;j++){
            cin >>arr[i][j];
        }
        tmp++;

    }
    tmp =2;
    t[0][0] = arr[0][0];
    for(int i=1;i<n;i++){
        for(int j =0;j<tmp;j++){
            if(j==0){
                t[i][j] = t[i-1][j]+arr[i][j];
            }
            else if(j==tmp-1){
                t[i][j] = t[i-1][j-1]+arr[i][j];
            }
            else{
                t[i][j] += max(t[i-1][j-1],t[i-1][j])+arr[i][j];
            }
        }
        tmp++;

    }
    int m = t[n-1][0];
    for(int i=1;i<n;i++){
        m = max(m,t[n-1][i]);
    }
    cout <<m;
    return 0;
}

File: java/baekjoon/src/_1932.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
public class _1932 {
    public static int N;
    public static int[][] dp;
    public static int[][] arr;
    public static int answer;
    public static void main(String[] args) throws IOException {
        BufferedReader br =new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        dp = new int[N+1][N+1];
        arr = new int[N+1][N+1];
        for(int i=1;i<=N;i++){
            StringTokenizer st = new StringTokenizer(br.readLine());
            for(int j=1;j<=i;j++) {
                int tmp =Integer.parseInt(st.nextToken());
                arr[i][j] = tmp;
            }
        }
        dp[1][1] = arr[1][1];
        for(int i=1;i<=N;i++){
            for(int j=1;j<=N;j++){
                dp[i][j] = Math.max(dp[i-1][j-1],dp[i-1][j])+arr[i][j];
                answer = Math.max(dp[i][j],answer);
            }
        }
//        for(int i=1;i<=N;i++){
//            for(int j=1;j<=i;j++){
//                System.out.print(dp[i][j]+" ");
//            }
//            System.out.println();
//        }
        System.out.println(answer);
    }
}
